materialize-bigquery/bigquery.go

```diff
@@ -1,214 +1,87 @@
+	for _, p := range projections {
+		if p.Field == fieldName {
+			projection = &p
+			break
+		}
 	}
-}
 
-// Bigquery only allows underscore, letters, numbers, and sometimes hyphens for identifiers. Convert everything else to underscore.
-var identifierSanitizerRegexp = regexp.MustCompile(`[^\-\._0-9a-zA-Z]`)
-
-func identifierSanitizer(text string) string {
-	return identifierSanitizerRegexp.ReplaceAllString(text, "_")
-}
-
-// SQLGenerator returns a SQLGenerator for the BigQuery SQL dialect.
-func SQLGenerator() sqlDriver.Generator {
-	var jsonMapper = sqlDriver.ConstColumnType{
-		SQLType: "STRING",
-		ValueConverter: func(i interface{}) (interface{}, error) {
-			switch ii := i.(type) {
-			case []byte:
-				return string(ii), nil
-			case json.RawMessage:
-				return string(ii), nil
-			case nil:
-				return json.RawMessage(nil), nil
-			default:
-				return nil, fmt.Errorf("invalid type %#v for variant", i)
-			}
-		},
+	if projection == nil {
+		return nil, errNotFound
 	}
 
-	var typeMappings = sqlDriver.ColumnTypeMapper{
-		sqlDriver.ARRAY:   jsonMapper,
-		sqlDriver.BINARY:  sqlDriver.RawConstColumnType("BYTES"),
-		sqlDriver.BOOLEAN: sqlDriver.RawConstColumnType("BOOL"),
-		sqlDriver.INTEGER: sqlDriver.RawConstColumnType("INT64"),
-		sqlDriver.NUMBER:  sqlDriver.RawConstColumnType("BIGNUMERIC"),
-		sqlDriver.OBJECT:  jsonMapper,
-		sqlDriver.STRING: sqlDriver.StringTypeMapping{
-			Default: sqlDriver.RawConstColumnType("STRING"),
-			ByFormat: map[string]sqlDriver.TypeMapper{
-				"date":      sqlDriver.RawConstColumnType("DATE"),
-				"date-time": sqlDriver.RawConstColumnType("TIMESTAMP"),
-			},
-		},
+	if projection.Inference.Exists == pf.Inference_INVALID || projection.Inference.Exists == pf.Inference_CANNOT {
+		return nil, errNotFound
 	}
 
-	var nullable sqlDriver.TypeMapper = sqlDriver.NullableTypeMapping{
-		NotNullText: "NOT NULL",
-		Inner:       typeMappings,
-	}
+	return &bigquery.FieldSchema{
+		Name:        projection.Field,
+		Type:        preferredFieldType(projection.Inference.Types),
+		Required:    projection.Inference.Exists == pf.Inference_MUST,
+		Description: projection.Inference.Description,
+	}, nil
+}
 
-	return sqlDriver.Generator{
-		Placeholder:        sqlDriver.QuestionMarkPlaceholder,
-		CommentRenderer:    sqlDriver.LineCommentRenderer(),
-		IdentifierRenderer: sqlDriver.NewRenderer(identifierSanitizer, sqlDriver.BackticksWrapper(), nil),
-		ValueRenderer:      sqlDriver.NewRenderer(sqlDriver.DefaultQuoteSanitizer, sqlDriver.SingleQuotesWrapper(), nil),
-		TypeMappings:       nullable,
+func preferredFieldType(possibleFields []string) bigquery.FieldType {
+	switch possibleFields[0] {
+	case "binary":
+		return bigquery.BytesFieldType
+	case "string":
+		return bigquery.StringFieldType
+	case "integer":
+		return bigquery.IntegerFieldType
+	case "number":
+		return bigquery.BigNumericFieldType
+	case "timestamp":
+		return bigquery.TimestampFieldType
+	default:
+		return bigquery.BytesFieldType
 	}
 }
```

Comment:
I'm not sure how useful/stable this method is, I think that is one area where I'm kinda flying blind. Would there be an example that I could use to validate must types?

Reply:
I think it may be better to panic here if we see a type that we do not recognise and can't translate to BigQuery directly. See the approach here:

https://github.com/estuary/flow/blob/master/crates/schemalate/src/firebolt/firebolt_schema_builder.rs#L160-L179
https://github.com/estuary/flow/blob/master/crates/schemalate/src/firebolt/firebolt_schema_builder.rs#L77-L80